### PR TITLE
refine mem allocator

### PR DESCRIPTION
**Pull request by ghost**
1.use a simpler allocator which uses mmap to search a proper place to allocate MemChunk(instead of read /proc/self/maps to guess the free address space)
2.use jmp addr64 to replace (push addr;ret;) ,which improves performance

### CODE REVIEW

File: inline_hook.cpp [hunk 1]
```diff
@@ -84,6 +87,17 @@ void GenerateJmpLarge(char* pWriteTo, void* pTarget)
 	pWriteTo += 4;
 	*pWriteTo = 0xc3;
 }
+/*
+NOTE: use long far jmp
+jmp addr64
+ff 25 00 00 00 00 addr
+ */
+void GenerateJmpLarge(char* pWriteTo, void* pTarget)
```
Comment (Menooker) [inline inline_hook.cpp:74]: This assembly seem not correct. I tried the bytes `ff 25 00 00 00 00, 01 02 03 04 05 06 07 08`, which should be `jmp 0x0102030405060708', but it turns out to be:

menooker@DESKTOP:/tmp$ echo 'asm(".byte 0xff, 0x25, 0x00,0x00,0x00,0x00,0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08");' > test.c; gcc -c test.c ;objdump -d test.o

test.o:     file format elf64-x86-64


Disassembly of section .text:

0000000000000000 <.text>:
   0:   ff 25 00 00 00 00       jmpq   *0x0(%rip)        # 0x6
   6:   01 02                   add    %eax,(%rdx)
   8:   03                      .byte 0x3
   9:   04 05                   add    $0x5,%al
   b:   06                      (bad)
   c:   07                      (bad)
   d:   08                      .byte 0x8

There should be no jmp absolute in x64. See reply `There is no jump of the form JMP absaddr to an absolute address in 64 bit mode. ` in [StackOverflow](https://stackoverflow.com/a/9819716)

File: inline_hook.cpp [hunk 2]
```diff
@@ -84,6 +87,17 @@ void GenerateJmpLarge(char* pWriteTo, void* pTarget)
...
+NOTE: use long far jmp
+jmp addr64
+ff 25 00 00 00 00 addr
+ */
+void GenerateJmpLarge(char* pWriteTo, void* pTarget)
+{
+	const char jmpl[]="\xff\x25\x00\x00\x00\x00";
+	memcpy(pWriteTo,jmpl,6);
+	memcpy(pWriteTo+6,&pTarget,8);
+}
```
Comment (Menooker) [inline inline_hook.cpp:79]: If you change this function, please change GetJmpLen as well. Thx

File: inline_hook.cpp [hunk 3]
```diff
@@ -325,6 +248,15 @@ inline bool InsertPatchPoint(PatchInfo* pool, int& num_patch_points, PatchType t
 	}\
 }while (0);
 
+
+static inline HookStatus rangeProtect(void* start,size_t size,int wrable){
```
Comment (Menooker) [inline inline_hook.cpp]: Maybe rangeProtect -> RangeProtect ?

File: inline_hook.cpp [hunk 4]
```diff
@@ -25,6 +24,10 @@ inline void* AlignToPage(void* addr)
 {
 	return (void*)((uintptr_t)addr & ~(PageSize2 - 1));
 }
+inline void* AlignToPage_UP(void* addr)
```
Comment (Menooker) [inline inline_hook.cpp:29]: Could you plz tell me what is this function for?

**Comment by Menooker** [comment]
I have an unrelated question: I am considering the compatibility of this library with Windows.  If I port this library to Windows, will it be more useful for you? :)

**Comment by Sysca11** [comment]
Sorry,I misunderstand this instruction. the correct explanation is "jmpq *0x0(rip);QWORD (the address to the newfunc)"------------------ 原始邮件 ------------------
发件人: "Menooker"<notifications@github.com>
发送时间: 2019年9月7日(星期六) 下午3:13
收件人: "Menooker/PFishHook"<PFishHook@noreply.github.com>;
抄送: "Sysca11"<3321056730@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [Menooker/PFishHook] refine mem allocator (#6)



@Menooker commented on this pull request.
 
 
In inline_hook.cpp:
 > @@ -84,6 +87,17 @@ void GenerateJmpLarge(char* pWriteTo, void* pTarget)  	pWriteTo += 4;  	*pWriteTo = 0xc3;  } +/* +NOTE: use long far jmp +jmp addr64 +ff 25 00 00 00 00 addr + */ +void GenerateJmpLarge(char* pWriteTo, void* pTarget)  
This assembly seem not correct. I tried the bytes ff 25 00 00 00 00, 01 02 03 04 05 06 07 08, which should be `jmp 0x0102030405060708', but it turns out to be:
 
menooker@DESKTOP:/tmp$ echo 'asm(".byte 0xff, 0x25, 0x00,0x00,0x00,0x00,0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08");' > test.c; gcc -c test.c ;objdump -d test.o
 
test.o:     file format elf64-x86-64
 
Disassembly of section .text:
 
0000000000000000 <.text>:
 0:   ff 25 00 00 00 00       jmpq   *0x0(%rip)        # 0x6
 6:   01 02                   add    %eax,(%rdx)
 8:   03                      .byte 0x3
 9:   04 05                   add    $0x5,%al
 b:   06                      (bad)
 c:   07                      (bad)
 d:   08                      .byte 0x8
 
There should be no jmp absolute in x64. See reply There is no jump of the form JMP absaddr to an absolute address in 64 bit mode.  in StackOverflow
 
&mdash;

**Comment by ghost** [comment]
> I have an unrelated question: I am considering the compatibility of this library with Windows. If I port this library to Windows, will it be more useful for you? :)

That's kind of you :D,but I'm not going to develop programs on Windows.
Thanks,anyway.

File: inline_hook.cpp (hunk 1 above)
Comment (ghost) [reply inline_hook.cpp:74]: jmpq QWORD PTR[rip];
.QWORD newfunc address

Comment (Menooker) [reply inline_hook.cpp:74]: Okay, I get it. It's an indirect jmp. Have you tried and checked this set of instructions?

**Comment by ghost** [comment]
<pre>inline int isHooked(void* oldfunc){
	return memcmp(oldfunc,"\xff\x25\x00\x00\x00\x00",6)==0;
}
HookStatus DupHook(void* oldfunc, void** poutold, void* newfunc){
	char* fc=AllocFunc(14,nullptr);
	memcpy(fc,oldfunc,14);
	*poutold=fc;
	if(rangeProtect(oldfunc,14,1)!=FHSuccess) return FHMprotectFail;
	memcpy((char*)oldfunc+6,&newfunc,8);
	if(rangeProtect(oldfunc,14,0)!=FHSuccess) return FHMprotectFail;
	return FHSuccess;
}
HookStatus HookIt(void* oldfunc, void** poutold, void* newfunc)
{
	if(isHooked(oldfunc)){
		return DupHook(oldfunc,poutold,newfunc);
	}</pre>

seems that HookIt cannot process "jmp qword ptr [rip];.QWORD".Hook a function twice will cause unexpected behavior

**Review by Menooker — COMMENTED** [review]
> inline int isHooked(void* oldfunc){
> 	return memcmp(oldfunc,"\xff\x25\x00\x00\x00\x00",6)==0;
> }
> HookStatus DupHook(void* oldfunc, void** poutold, void* newfunc){
> 	char* fc=AllocFunc(14,nullptr);
> 	memcpy(fc,oldfunc,14);
> 	*poutold=fc;
> 	if(rangeProtect(oldfunc,14,1)!=FHSuccess) return FHMprotectFail;
> 	memcpy((char*)oldfunc+6,&newfunc,8);
> 	if(rangeProtect(oldfunc,14,0)!=FHSuccess) return FHMprotectFail;
> 	return FHSuccess;
> }
> HookStatus HookIt(void* oldfunc, void** poutold, void* newfunc)
> {
> 	if(isHooked(oldfunc)){
> 		return DupHook(oldfunc,poutold,newfunc);
> 	}
> seems that HookIt cannot process "jmp qword ptr [rip];.QWORD".Hook a function twice will cause unexpected behavior

Yes, but I suppose hooking a function twice is not allowed, so we can just ignore the case. And the problem hooking "jmp qword ptr [rip];.QWORD" is that the disassembler cannot process the `DWORD` part, since it is not an instruction. It is pure data. 

But I think we can keep your `isHooked` and `DupHook` as a special case handler.

BTW, the naming convension is supposed to be `SomeFunction` not `someFunction`. :)

**Comment by Menooker** [comment]
Please view my comments, and maybe polish the code. Thanks a lot!

File: inline_hook.cpp (hunk 2 above)
Comment (Menooker) [reply inline_hook.cpp:79]: Oh... The new jmp and the old jmp assembly has the same size, right? No need to change this?

**Comment by ghost** [comment]
I think patch instructions which use rip-relative jump,and make sure the instructions which use rip-relative memory read/write have the correct memory address is enough.
For example,jmp rip+0x18 should be patched to jmp rip+xxx,but jmp qword ptr [rip+0] should not be patched.copy the next qword to shadow function is a good choice.
This means if we meet a instruction needs the memory at [oldfunc,oldfunc+14],we shouldn't patch the rip-offset.Instead,copy it directly,and also copy the memory it accesses

File: inline_hook.cpp (hunk 4 above)
Comment (ghost) [reply inline_hook.cpp:29]: align to the ceil page

File: inline_hook.cpp (hunk 3 above)
Comment (ghost) [reply inline_hook.cpp]: changed

File: inline_hook.cpp (hunk 2 above)
Comment (ghost) [reply inline_hook.cpp:79]: size=6+8=14

File: inline_hook.cpp (hunk 1 above)
Comment (ghost) [reply inline_hook.cpp:74]: checked

**Comment by Menooker** [comment]
I think you are right. we should patch rip+0 as a special case

**Pull request merged**